src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -999,9 +1025,29 @@ void IHttp2StreamLifetimeHandler.OnStreamCompleted(int streamId)
         {
             lock (_stateLock)
             {
-                _streams.TryRemove(streamId, out _);
+                // Get, Add, Remove so the steam is always registered in at least one collection at a time.
+                if (_streams.TryGetValue(streamId, out var stream))
+                {
+                    _activeStreamCount--;
+
+                    if (stream.IsDraining)
+                    {
+                        stream.DrainExpiration =
+                            _context.ServiceContext.SystemClock.UtcNow + Constants.RequestBodyDrainTimeout;
 
-                if (_streams.IsEmpty)
+                        _drainingStreams.TryAdd(streamId, stream);
+                    }
+                    else
```

Comment:
When would this happen? Doesn't OnRequestProcessingEnded always set the IsDraining flag before calling OnStreamCompleted?

Reply:
No, a stream only goes into draining mode if EndOfStream or Reset have not already been received.